arwen/elrondapi/managedei.go

```diff
@@ -0,0 +1,1035 @@
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.GetOwnerAddress
+	metering.UseGas(gasToUse)
+
+	owner, err := blockchain.GetOwnerAddress()
+	if arwen.WithFault(err, context, runtime.ElrondAPIErrorShouldFailExecution()) {
+		return
+	}
+
+	managedType.SetBytes(destinationHandle, owner)
+}
+
+//export v1_4_managedCaller
+func v1_4_managedCaller(context unsafe.Pointer, destinationHandle int32) {
+	managedType := arwen.GetManagedTypesContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.GetCaller
+	metering.UseGas(gasToUse)
+
+	caller := runtime.GetVMInput().CallerAddr
+	managedType.SetBytes(destinationHandle, caller)
+}
+
+//export v1_4_managedSignalError
+func v1_4_managedSignalError(context unsafe.Pointer, errHandle int32) {
+	managedType := arwen.GetManagedTypesContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.SignalError
+	metering.UseGas(gasToUse)
+
+	errBytes, err := managedType.GetBytes(errHandle)
+	if arwen.WithFault(err, context, runtime.ManagedBufferAPIErrorShouldFailExecution()) {
+		return
+	}
+	managedType.ConsumeGasForThisIntNumberOfBytes(len(errBytes))
+
+	runtime.SignalUserError(string(errBytes))
+}
+
+//export v1_4_managedWriteLog
+func v1_4_managedWriteLog(
+	context unsafe.Pointer,
+	topicsHandle int32,
+	dataHandle int32,
+) {
+	runtime := arwen.GetRuntimeContext(context)
+	output := arwen.GetOutputContext(context)
+	metering := arwen.GetMeteringContext(context)
+	managedType := arwen.GetManagedTypesContext(context)
+
+	topics, sumOfTopicByteLengths, err := readManagedVecOfManagedBuffers(managedType, topicsHandle)
+	if arwen.WithFault(err, context, runtime.ElrondAPIErrorShouldFailExecution()) {
+		return
+	}
+
+	dataBytes, err := managedType.GetBytes(dataHandle)
+	if arwen.WithFault(err, context, runtime.ManagedBufferAPIErrorShouldFailExecution()) {
+		return
+	}
+	managedType.ConsumeGasForThisIntNumberOfBytes(len(dataBytes))
+	dataByteLen := uint64(len(dataBytes))
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.Log
+	gasForData := math.MulUint64(
+		metering.GasSchedule().BaseOperationCost.DataCopyPerByte,
+		sumOfTopicByteLengths+dataByteLen)
+	gasToUse = math.AddUint64(gasToUse, gasForData)
+	metering.UseGas(gasToUse)
+
+	output.WriteLog(runtime.GetSCAddress(), topics, dataBytes)
+}
+
+//export v1_4_managedGetOriginalTxHash
+func v1_4_managedGetOriginalTxHash(context unsafe.Pointer, resultHandle int32) {
+	runtime := arwen.GetRuntimeContext(context)
+	metering := arwen.GetMeteringContext(context)
+	managedType := arwen.GetManagedTypesContext(context)
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.GetBlockHash
+	metering.UseGas(gasToUse)
+
+	managedType.SetBytes(resultHandle, runtime.GetOriginalTxHash())
```

Comment:
Luckily, `RuntimeContext.SetVMInput()` copies these bytes already, so they should be safe.